### PR TITLE
Port #966 and #414 from yuzu (Cleanups in common and file_util) & Make bitfield assignment operator public

### PR DESCRIPTION
**Pull request by FearlessTobi**
Please see yuzu-emu/yuzu#966 and yuzu-emu/yuzu#414 for more details.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4132)

### CODE REVIEW

File: src/common/file_util.h [hunk 1]
```diff
@@ -174,13 +174,9 @@ class IOFile : public NonCopyable {
     bool Close();
 
     template <typename T>
-    size_t ReadArray(T* data, size_t length) {
```
Comment (MerryMage) [inline src/common/file_util.h]: You cannot change this to be a `const` member function.

File: src/common/file_util.h [hunk 2]
```diff
@@ -215,17 +207,21 @@ class IOFile : public NonCopyable {
         return items_written;
     }
 
-    size_t ReadBytes(void* data, size_t length) {
+    template <typename T>
+    size_t ReadBytes(T* data, size_t length) const {
```
Comment (MerryMage) [inline src/common/file_util.h]: Ditto.

File: src/common/bit_field.h [hunk 3]
```diff
@@ -128,6 +122,8 @@ struct BitField {
     using StorageTypeU = std::make_unsigned_t<StorageType>;
 
 public:
+    BitField& operator=(const BitField&) = default;
```
Comment (wwylele) [inline src/common/bit_field.h]: Where does this come from? Although I wanted this change, but it deserves its own PR if it is intended.

Comment (wwylele) [reply src/common/bit_field.h]: Could you change the PR title & commit description and add comment specifically for this? It wasn't in private for no reason, as it was documented in the old code. Please understand that putting this to public, while being required by some operation, is dangerous for other potential mistake. Dangerous code change needs a good marking in git history.
